crates/sdl-encoder/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+
+    /// Adds a new Schema Definition.
+    ///
+    /// The schema type is only used when the root GraphQL type is different
+    /// from default GraphQL types.
+    pub fn schema(&mut self, schema: SchemaDef) {
+        self.buf.push_str(&schema.to_string());
+    }
+
+    /// Return the encoded SDL string after all types have been
+    pub fn finish(self) -> String {
+        self.buf
+    }
+}
+
+impl Default for Schema {
+    fn default() -> Self {
+        Schema::new()
+    }
+}
+
+/// Type Definition used to define different types in SDL.
+#[derive(Debug)]
+pub struct TypeDef {
+    name: String,
+    description: Option<String>,
+    fields: Vec<Field>,
+}
+
+impl TypeDef {
+    /// Create a new instance of TypeDef with a name.
+    pub fn new(name: String, field: Field) -> Self {
+        Self {
+            name,
+            description: None,
+            fields: vec![field],
+        }
+    }
+
+    /// Set the TypeDef's description field.
+    pub fn description(&mut self, description: String) {
+        self.description = Some(description)
+    }
+
+    /// Push a Field to type def's fields vector.
+    pub fn field(&mut self, field: Field) {
+        self.fields.push(field)
+    }
+}
+
+impl Display for TypeDef {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(description) = &self.description {
+            writeln!(f, "\"\"\"\n{}\n\"\"\"", description)?;
+        }
+
+        let mut fields = String::new();
+        for field in &self.fields {
+            fields += &format!("\n{}", field.to_string());
+        }
+
+        write!(f, "type {} {{", &self.name)?;
+        write!(f, "{}", fields)?;
+        writeln!(f, "\n}}")
+    }
+}
+
+/// A definition used when a root GraphQL type differs from default types.
+#[derive(Debug)]
+pub struct SchemaDef {
+    description: Option<String>,
+    fields: Vec<Field>,
+}
+
+impl SchemaDef {
+    /// Create a new instance of SchemaDef.
+    pub fn new(field: Field) -> Self {
+        Self {
+            description: None,
+            fields: vec![field],
+        }
+    }
+
+    /// Set the schema def's description.
+    pub fn description(&mut self, description: String) {
+        self.description = Some(description)
+    }
+
+    /// Push a Field to schema def's fields vector.
+    pub fn field(&mut self, field: Field) {
+        self.fields.push(field)
+    }
+}
+
+impl Display for SchemaDef {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(description) = &self.description {
+            writeln!(f, "\"\"\"\n{}\n\"\"\"", description)?;
+        }
+
+        let mut fields = String::new();
+        for field in &self.fields {
+            fields += &format!("\n{}", field.to_string());
+        }
+
+        write!(f, "schema {{")?;
+        write!(f, "{}", fields)?;
+        writeln!(f, "\n}}")
+    }
+}
+
+/// Define whether a Field is a vector.
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum VecState {
```

Comment:
This won't work for nested lists (e.g. `[[String]]`), and it also won't allow you to differentiate between nullable and non-nullable element types (e.g. `[String]` vs `[String!]`). That would be a good reason to use a more structured representation (probably an enum, like we [do in Apollo iOS](https://github.com/apollographql/apollo-ios/blob/cd9a77bac40c61f547a6c942bca2fcd74114d0d5/Sources/ApolloCodegenLib/Frontend/GraphQLType.swift#L4)).

Reply:
Ah yes, this does not account for `[[String]]`! the `[String]` and `[String!]` is accounted for within the Field type under the `nullable` property. I'll think about a way to accommodate lists though.

Reply:
Note that there is a difference between `[String!]` and `[String]!]` though (and `[String!]!]`, and `[[String]!]`, etc.). I really think this requires a recursive type definition to get it right.

Reply:
yes, of course the parsed introspection result will have to be read recursively. But that's not part of the serialiser's job -- that will be handled in a different part of an application. This module literally only cares about creating the string.

Reply:
But what I mean is that in order to generate the correct string you need to be able to differentiate between these different types. It isn't enough to know that a field is nullable, you need to know if the element type of a list (if it is a list) is nullable, or even several layers down in the case of nested lists. This is how we generate strings for type references in Apollo iOS: https://github.com/apollographql/apollo-ios/blob/cd9a77bac40c61f547a6c942bca2fcd74114d0d5/Sources/ApolloCodegenLib/Frontend/GraphQLType.swift#L9

Reply:
Ah I see what you mean! I think I have a good way to fix this.